### PR TITLE
CFEP-01 CFEP Purpose and Guidelines and CFEP-00 CFEP Template

### PR DESCRIPTION
**Pull request by jjhelmus**
Conda-Forge Enhancement Proposal for CFEPs themselves as well as a template for
future CFEPs.

### CODE REVIEW

File: cfep-01.md
Comment (jjhelmus) [inline cfep-01.md]: The exactly mechanism on how a consensus is reached should be decided upon and included here.

File: cfep-01.md
Comment (jjhelmus) [inline cfep-01.md]: How long an _appropiate amount of time_ is will depend on the issue and its discussion.  Should something about this be added here?

**Comment by jjhelmus** [comment]
Rendered versions of the README and CFEPs can be found in [the jjhelmus/cfep01 branch](https://github.com/jjhelmus/conda-forge-enhancement-proposals/tree/cfep01)

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md:78]: A little nervous about the _Deferred_ status. For instance, what limits how long something can be deferred? How do we ensure we actually reach a decision? Right now there are things that have effectively been deferred indefinitely. If we keep it, there need to be some limits on this so it actually gets accepted or rejected at some point. In some ways, I'd much rather see a proposal outright rejected if that is the case so that we can start making other plans.

Note that JEPs appear to exclude this status from the [ones available](https://github.com/jupyter/enhancement-proposals/blob/dd150b31ab1eddc7a2409205bfd40b2560027fb3/jupyter-enhancement-proposal-guidelines/jupyter-enhancement-proposal-guidelines.md#jep-labels) too.

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md]: One common theme I'm starting to pick up on is providing the ability for people to transfer ownership. This seems to exist for many reasons including, but not limited to, not having time to champion it further. For instance, the [first PEP](https://www.python.org/dev/peps/pep-0001/#transferring-pep-ownership) lays this out nicely. Similarly JEP's have a way to [withdraw](https://github.com/jupyter/enhancement-proposals/blob/dd150b31ab1eddc7a2409205bfd40b2560027fb3/jupyter-enhancement-proposal-guidelines/jupyter-enhancement-proposal-guidelines.md#jep-labels) when it is no longer actively supported. I can see a lot of value in us also supporting this mode. However, that would actually change the model proposed here. Namely we would merge CFEPs in draft mode and then continue discussing. Would be curious to hear your thoughts on this and the thoughts of others in the community.

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md]: nit: Space needed between `)` and `which`.

File: cfep-00.md
Comment (jakirkham) [inline cfep-00.md]: There is a nice layout below that briefly explains the purpose of each of these sections below. It might be worthwhile to include in the template so people can see it when writing their proposal. Then again it could be noisy. Would like to hear people's thoughts on this generally.

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md:86]: This may seem like a small point, but actually I can see value in doing this as a second step. Namely we merge in _Draft_ mode and then later someone (not even necessarily the original author) updates this to rejected. The value being twofold.
1. We are not relying on someone that just had their CFEP rejected to do this work. This seems minor, but I can certainly see people leaving abandoned PRs that were rejected.
2. We can offer up a summary judgement. Why did we reject it. What changes might allow for a future proposal to be accepted. Etc.

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md:121]: One field that might be nice, but is not necessarily filled out by the author would be a Resolution field. Basically what did we do with this CFEP and why.

This goes beyond linking the PR(s) that implements this functionality, but includes why did we accept it or why did we reject. These summary opinions can be very helpful for historic record, proposing new CFEPs in the same vein, or generally getting a feel for what makes for a working CFEP. PEPs do this with a [link to a comment](https://www.python.org/dev/peps/pep-0001/#pep-header-preamble). Though they also use a mailing list. GitHub comments as they are with the ability to edit in a quasi-transparent manner may not be the best thing to link to. Instead maybe a couple line summary that goes in with the decision would be nice.

File: cfep-00.md
Comment (jakirkham) [inline cfep-00.md]: This halfway between question and suggestion. Should we leave this open ended or should we just stick them all under one license?

File: cfep-01.md
Comment (jakirkham) [inline cfep-01.md]: Probably should have an author(s) field too.

**Comment by jakirkham** [comment]
Thanks for putting this together @jjhelmus. This is very nicely done.

Basically my comments fit along one line, which is inspired based on what @pelson had said in our meeting discussing governance and CFEPs. Namely that it should be ok to merge things in Draft mode and often. Only after that do we start digging into a discussion about the proposal and come to some decision.

Other comments are directed at the decision process and making sure that it provides clear feedback and does happen.

**Comment by jakirkham** [comment]
@conda-forge/core, please take a look at these documents that @jjhelmus has nicely put together for us and provide constructive feedback. These are laying the groundwork for how we make major decision on large changes going forward. So become familiar with it, ask questions about unclear portions, and make suggestions. We want to be sure we have firm footing going forward.

File: cfep-01.md
Comment (patricksnape) [reply cfep-01.md:86]: I agree that both of those points seem pragmatic.

File: cfep-00.md
Comment (patricksnape) [reply cfep-00.md]: I say we enforce a public license in order not to have a CFEP submitted that just devolves into discussing what license it was submitted under.

File: README.md
Comment (mwcraig) [inline README.md]: `community led` → `community-led`

File: cfep-00.md
Comment (mwcraig) [inline cfep-00.md]: This will be a link to the _pull request_ for the proposal, not an issue, right?

File: cfep-00.md
Comment (mwcraig) [reply cfep-00.md]: How about adding a link here to CFEP-01? I like the simplicity of this template but a convenient link to more detail would be helpful.

File: cfep-00.md
Comment (mwcraig) [reply cfep-00.md]: Suggested wording: `All CFEPs are explicitly public domain.`

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md]: Could you please add a link to the hackpad for the meetings? Otherwise there is no way for someone to find out  about the agendas, times, etc.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md:41]: I think the mailing list is a better forum for this kind of question because it is easier to find the discussion later. gitter is pretty stream-of-conciousness.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md]: `,` → `;`

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md:50]: Suggestion 1: drop, because it is in the previous paragraph.
Suggestion 2: if this sentence is kept, put the mailing list before gitter.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md:56]: I'm don't think this pre-clearance is necessary. If an idea is proposed that has many more downsides than upsides then the CFEP process will proceed quickly and result in rejecting the CFEP. The benefit is that the reasons for rejection will be clearly documented and if the issue arises again we have something concrete and easy-to-find to point people to.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md]: This can be dropped. I think we can assume that champions will get feedback on their draft before making the formal submission, so permission isn't needed, and I would like to funnel feedback through CFEP pull requests rather than through informal channels.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md:46]: What about changing "champion" to "advocate"?

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md]: Perhaps replace it with this line:

"While working on the CFEP prior to submission please set its Status to Draft."

File: cfep-01.md
Comment (mwcraig) [reply cfep-01.md:86]: It seems reasonable to expect that one of the core conda-forge maintainers would handle these updates regardless of whether the CFEP is rejected, accepted, or deferred. The mechanism for doing this (a PR on the PR to the CFEP repo, or a merge followed by a PR esolving the status) doesn't seem that important. 

I'm fine with the latter, which is what I think @jakirkham is proposing, in the interests of keeping these moving along once they are resolved.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md:92]: Suggestion: For all CFEPs, regardless of final status, should be amended to include a link to the PR in which it was discussed.

File: cfep-01.md
Comment (mwcraig) [reply cfep-01.md]: Suggestion: "after an appropriate amount of time" → "if the circumstances that justified for deferring instead of accepting or rejecting have changed"

I'm fishing for something that would allow for a CFEP that is deferred because, e.g. it would require functionality in conda-build that is not yet implemented. It may be difficult to put a time horizon on that, but if this hypothetical new feature is implemented (or a decision is made to never implement it) then clearly there is no longer a reason to defer.

edit: added "of" so the suggested wording makes sense

File: cfep-01.md
Comment (mwcraig) [reply cfep-01.md:121]: I do think it is helpful to link to the github PR for the CFEP, even though comments can be edited.  Github does record whether edits have been made and when.

I do agree that a brief summary is useful regardless.

File: cfep-01.md
Comment (mwcraig) [inline cfep-01.md]: Is anything more concrete (e.g. CC-0) needed here? Am fine with it as is...

**Comment by mwcraig** [comment]
Thanks for putting this together @jjhelmus!

File: cfep-01.md
Comment (jakirkham) [reply cfep-01.md]: Yeah I think someone mentioned this elsewhere, but Public Domain doesn't exist everywhere so CC0 seems like a sensible choice that avoids these problems. That all said IANAL.

File: cfep-00.md
Comment (jakirkham) [reply cfep-00.md]: ...or CC0 as Public Domain is not honored everywhere. Again IANAL.

Comment (scopatz) [reply cfep-00.md]: Yes, please change to CC0 rather than public domain.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:78]: I like @mwcraig wording from above 

_Deferred_ CFEPs can be re-submitted in a new pull request if the circumstances that justified deferring instead of accepting or rejecting have changed.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:46]: "champion" is used in Python PEP, I like the word.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:92]: The PR where a CFEP is discussed should be linked to in table at the top of the CFEP.  Adjusted the text to emphasize this.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Agreed, accepted.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: I have no problem with the idea of transferring ownership of a CFEP and can add a paragraph about it if needed but I'm also alright keeping this informal for the time being.  Maybe this can be added to the CFEP if we find that a more formal mechanism is needed.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Done

File: cfep-00.md
Comment (jjhelmus) [reply cfep-00.md]: Copied the summaries from CFEP-01 to the template.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:86]: Changed the text to indicate that the status of the CFEP should be modified by conda-forge maintainers prior to merging. 

I'll comment on the two step vs one step workflow in the main thread so it does not get lost.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:121]: Added a Resolution section.

File: cfep-00.md
Comment (jjhelmus) [reply cfep-00.md]: Changed to CC0 1.0 Universal

File: README.md
Comment (jjhelmus) [reply README.md]: Done

File: cfep-00.md
Comment (jjhelmus) [reply cfep-00.md]: Yes it should be, changed.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Added.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:41]: Agreed.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Done

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:50]: Retained the sentence but removed the gitter link

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md:56]: This floating an idea was taken from Python PEPs workflow.  I'm happy to remove it if it seems unnecessary.  We can always add it back in if we find that too many CFEPs start pouring in.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Done.

File: cfep-01.md
Comment (jjhelmus) [reply cfep-01.md]: Changed to CC0

**Comment by jjhelmus** [comment]
@jakirkham suggested:

> This may seem like a small point, but actually I can see value in doing this as a second step. Namely we merge in Draft mode and then later someone (not even necessarily the original author) updates this to rejected. The value being twofold.
> 
> We are not relying on someone that just had their CFEP rejected to do this work. This seems minor, but I can certainly see people leaving abandoned PRs that were rejected.
> We can offer up a summary judgement. Why did we reject it. What changes might allow for a future proposal to be accepted. Etc.

I've given some thought to a single step proposal process vs a two step process.  Both have their merits and I'm really happy with either.  The reason I prefer the single step process is that it allows an author to easily update the CFEP based upon feedback from the community and these is a single location where all discussion of the proposal takes place.  

I think the mechanism for a two step process is more complicated.  The proposal must first be reviewed only for formatting issues.  I'm sure there would be many who would want to comment on the content which could result in the discussion being split into two locations.  Once the draft is accepted the workflow seems a bit odd.  Does the author open a new PR with no changed to the CFEP?  Open an issue?  If changes are requested how is this done?

I retained the single step workflow in the current draft but changed the text so that conda-forge maintainers change the status before merging the PR.

**Comment by jjhelmus** [comment]
Thanks everyone for the feedback.  I've tried to incorporate as many of the suggestions as possible into the new draft.

**Comment by jakirkham** [comment]
Sure, so part of the reason I mentioned this is Jupyter tried doing it in one PR, but ultimately found it better to do subsequent PRs. See this [discussion](https://github.com/jupyter/enhancement-proposals/pull/2#discussion_r76999888). In light of this [GitHub addition](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/), maybe it doesn't matter much.

I'm not picky either. Just trying to anticipate possible problems.

File: cfep-01.md
Comment (pelson) [inline cfep-01.md:100]: A github handle would do it too.

**Comment by pelson** [comment]
So here is an interesting one - how do we go from proposed to accepted? Merge then new PR, or do it in the PR?

**Review by pelson — APPROVED** [review]

**Pull request merged**
